### ISSUE TITLE
Flaky Test: CogRPCClientTest > deliver cargo and receive ack

### ISSUE THREAD

**Issue by sdsantos** [relaycorp/awala-cogrpc-jvm#282]
```
CogRPCClientTest > deliver cargo and receive ack() FAILED
    java.util.NoSuchElementException at CogRPCClientTest.kt:48
```

**Comment by gnarea** [comment]
Reopening because #292 didn't fix this. See:

- Merge into `master`: https://github.com/relaycorp/awala-cogrpc-jvm/actions/runs/4992151270/jobs/8939564488
- Another PR with the changes from #292: https://github.com/relaycorp/awala-cogrpc-jvm/actions/runs/4992154523/jobs/8939571696

cc: @Filmaluco

**Comment by sdsantos** [comment]
Investigated a bit more and adding here more information. Here what's going on on those 2 last failures, in terms of logs:

```
May 16, 2023 1:02:38 PM tech.relaycorp.relaynet.cogrpc.client.CogRPCClient$deliverCargo$1 invokeSuspend
INFO: deliverCargo next 0dc0c21a-8d0d-4086-ac77-2f58cf721dcb
May 16, 2023 1:02:44 PM tech.relaycorp.relaynet.cogrpc.client.CogRPCClient$deliverCargo$1$ackObserver$1 onError
WARNING: Ending deliverCargo due to ack error
io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 4.543277700s. [closed=[], open=[[buffered_nanos=5270378100, waiting_for_connection]]]
```

```
May 16, 2023 1:02:10 PM tech.relaycorp.relaynet.cogrpc.client.CogRPCClient$deliverCargo$1 invokeSuspend
INFO: deliverCargo next 4c82aa3b-53fe-44f2-9cdc-a628d24f81d1
May 16, 2023 1:02:10 PM tech.relaycorp.relaynet.cogrpc.client.CogRPCClient$deliverCargo$1$ackObserver$1 onNext
INFO: deliverCargo ack 4c82aa3b-53fe-44f2-9cdc-a628d24f81d1
May 16, 2023 1:02:15 PM tech.relaycorp.relaynet.cogrpc.client.CogRPCClient$deliverCargo$1$ackObserver$1 onNext
INFO: deliverCargo complete
May 16, 2023 1:02:15 PM tech.relaycorp.relaynet.cogrpc.client.CogRPCClient$deliverCargo$1$ackObserver$1 onError
WARNING: Ending deliverCargo due to ack error
io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 3.781601600s. [closed=[], committed=[remote_addr=localhost/127.0.0.1:8080]]
```

Basically, they are behaving as if the server never acked back the cargo received (the first one), or as if the server never closed its stream after receiving `onCompleted`. Both of those situations are weird because we're specifically setting this:

```
        mockServerService.deliverCargoReturned = object : NoopStreamObserver<CargoDelivery>() {
            override fun onNext(value: CargoDelivery) {
                mockServerService.deliverCargoReceived?.onNext(cargo.toCargoDelivery().toAck())
            }
            override fun onCompleted() {
                mockServerService.deliverCargoReceived?.onCompleted()
            }
        }
```

---

Before the actual test logs were also getting a bunch of errors like these which seem unrelated with us, but also GRPC, which is cryptic:

```
May 16, 2023 1:02:38 PM io.grpc.internal.ManagedChannelOrphanWrapper$ManagedChannelReference cleanQueue
SEVERE: *~*~*~ Previous channel ManagedChannelImpl{logId=5, target=directaddress:////192.168.43.1:80} was not shutdown properly!!! ~*~*~*
    Make sure to call shutdown()/shutdownNow() and wait until awaitTermination() returns true.
java.lang.RuntimeException: ManagedChannel allocation site
```

But maybe they are the clue of things going wrong only in the java-17-windows machine?

---

Theories:

- Maybe the tests are running in parallel and there are multiple servers being setup in the same port, with different behaviors? The failing test is always the first, and it requires specific behavior from the server (acking and completing) instead of the default NoOp we had. Maybe that second test is overriding the behavior halfway?
- Maybe the server is being killed too soon for some reason?

I'll probably try next randomizing the port for each test and adding some extra logs.

**Comment by gnarea** [comment]
Thanks for looking into this Sergio!

Your assessment sounds about right to me. Did you consider if the following may be the root cause?

https://github.com/relaycorp/awala-cogrpc-jvm/blob/c11358d4fe46adcc7fbbdee7143dcbfde76d1aca/src/test/resources/junit-platform.properties#L1
